File: M3-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        windowStr = ''
        maxLength = 0
        for item in s:
            previousStr = windowStr
            windowStr += item
            if item in previousStr:
                itemIndex = windowStr.index(item)
                windowStr = windowStr[itemIndex+1:len(windowStr)]
            maxLength = max(maxLength,len(windowStr))
        return maxLength

File: Java/M3-longest-substring-without-repeating-characters.java

import java.lang.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int maxlength = 0;
        int left = 0;
        for(int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            if(map.containsKey(c)) {
                left = Math.max(left, map.get(c) + 1);
            }
            maxlength = Math.max(i - left + 1, maxlength);
            map.put(c, i);
        }
        return maxlength;
        
    }
}